### PR TITLE
TL/MLX5: MCAST HCA selection and COLL score

### PR DESCRIPTION
**Pull request by MamziB**
TL/MLX5: MCAST HCA selection and COLL score

### CODE REVIEW

File: src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c [hunk 1]
```diff
@@ -50,6 +47,9 @@ ucc_status_t ucc_tl_mlx5_mcast_context_init(ucc_tl_mlx5_mcast_context_t    *cont
     int                     user_provided_ib;
     int                     ib_valid;
     const char             *dst;
+    char                   *ib_devname = NULL;
```
Comment (samnordmann) [inline src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:50]: Please place initialized variables before uninitialized

File: src/components/tl/mlx5/tl_mlx5.h [hunk 2]
```diff
@@ -17,7 +17,7 @@
 #include "mcast/tl_mlx5_mcast.h"
 
 #ifndef UCC_TL_MLX5_DEFAULT_SCORE
-#define UCC_TL_MLX5_DEFAULT_SCORE 1
```
Comment (samnordmann) [inline src/components/tl/mlx5/tl_mlx5.h:20]: copyright

File: src/components/tl/mlx5/tl_mlx5_coll.c [hunk 3]
```diff
@@ -51,6 +51,7 @@ ucc_status_t ucc_tl_mlx5_task_finalize(ucc_coll_task_t *coll_task)
     if (req != NULL) {
         ucc_assert(coll_task->status != UCC_INPROGRESS);
         ucc_free(req);
+        tl_trace(UCC_TASK_LIB(task), "finalizing an mcast task %p", task);
```
Comment (samnordmann) [inline src/components/tl/mlx5/tl_mlx5_coll.c:54]: copyright

**Review by samnordmann — COMMENTED** [review]
Looks good!
About the score,
I think that we should have a score of 0 if mcast is disabled here: https://github.com/openucx/ucc/blob/f01fd27e0452052be76454d995c1b60497a4782f/src/components/tl/mlx5/tl_mlx5_team.c#L226

File: src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c [hunk 4]
```diff
@@ -87,11 +87,33 @@ ucc_status_t ucc_tl_mlx5_mcast_context_init(ucc_tl_mlx5_mcast_context_t    *cont
...
         strncat(ctx->devname, ":1", 3);
         user_provided_ib = 0;
+        ctx->ib_port     = 1;
     } else {
         ib_valid = 0;
         /* user has provided the devname now make sure it is valid */
+        /* check if port number is also included and extract devname from user str */
+        ib_devname  = mcast_ctx_conf->ib_dev_name;
+        pos         = strstr(ib_devname, ":");
+        end_pos     = ib_devname + strlen(ib_devname);
```
[the remark this answers is not in the document]
Comment (MamziB) [reply src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c:97]: you are right it is not needed, we need to also change the mlx5_context.c file as well, as this code block is copied from there.

**Comment by MamziB** [comment]
@Sergei-Lebedev @samnordmann  I resolved all the comments. Thanks for the reviewing this PR

**Review by samnordmann — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**